rfcs/text/0002_encrypted_attributes.md

```diff
@@ -0,0 +1,228 @@
+properties and passes that same object to the `update` Saved Objects API method without re-defining encrypted attributes.
+At this stage we consider this as a developer mistake and don't prevent it from happening in any way apart from logging
+this type of event.
+
+Saved object ID is an essential part of Additional authenticated data (AAD) used during encryption process and hence
+should be as hard to guess as possible. To fulfil this requirement wrapper generates highly random IDs (UUIDv4) for the
+saved objects that contain encrypted attributes and hence consumers are not allowed to specify ID when calling `create`
+or `bulkCreate` method and if they try to do so the error will be thrown.
+
+To reduce the risk of unintentional decryption and consequent leaking of the sensitive information there is only one way
+to retrieve saved object and decrypt its encrypted attributes and it's exposed only through `encrypted_saved_objects` plugin:
+
+```typescript
+const savedObject = await server.plugins.encrypted_saved_objects.getDecryptedAsInternalUser(
+  'server-action',
+  'dd9750b9-ef0a-444c-8405-4dfcc2e9d670'
+);
+
+// savedObject = { 
+//   id: 'dd9750b9-ef0a-444c-8405-4dfcc2e9d670',
+//   type: 'server-action',
+//   name: 'my-server-action',
+//   credentials: { username: 'some-user', password: 'some-password' },
+//   apiKey: 'dGhpcyBpcyBub3QgYSByZWFsIHRva2VuIGJ1dCBpdCBpcyBvb',
+// };
+```
+
+As can be seen from the method name, the request to retrieve saved object and decrypt its attributes is performed on
+behalf of the internal Kibana user and hence isn't supposed to be called within user request context.
+
+**Note:** the fact that saved object with encrypted attributes is created using standard Saved Objects API within a 
+particular user and space context, but retrieved out of any context makes it unclear how consumers are supposed to 
+provide that context and retrieve saved object from a particular space. Current plan for `getDecryptedAsInternalUser` 
+method is to accept a third `BaseOptions` argument that allows consumers to specify `namespace` that they can retrieve
+from the request using public `spaces` plugin API.
+
+## Encryption and decryption
+
+Saved object attributes are encrypted using [@elastic/node-crypto](https://github.com/elastic/node-crypto) library. Please
+take a look at the source code of this library to know how encryption is performed exactly, what algorithm and encryption 
+parameters are used, but in short it's AES Encryption with AES-256-GCM that uses random initialization vector and salt.
+
+As with encryption key for Kibana's session cookie, master encryption key used by `encrypted_saved_objects` plugin can be
+defined as a configuration value (`xpack.encrypted_saved_objects.encryptionKey`) via `kibana.yml`, but it's **highly 
+recommended** to define this key in the [Kibana Keystore](https://www.elastic.co/guide/en/kibana/current/secure-settings.html)
```

Comment:
Is it possible to determine when we're using the keystore from within a plugin? If so, we should log a warning whenever we're not using the keystore for this, and codify that behavior in this RFC.

Reply:
Well, not really, unless the core starts providing `source` (`env`, `config`, `keystore` or `cli`) with every config value.

We can potentially try to access key store directly from plugin and check whether value it has equals to the value we get from config, but we can't be 100% sure that actual config value came from keystore and not defined somewhere else (e.g. duplicated in yaml or env variable)... So I'm kind of on the fence here.

What do you think?

Reply:
I think it'd be great if we could implement this behavior, but it feels like this should be considered out of scope for this RFC. We already have quite a few settings which ideally should be configured using the keystore and not using the kibana.yml. Do you see the usage in this RFC differing enough that this would become a prerequisite @epixa?

Reply:
Works for me.